core/src/apps/bitcoin/scripts_decred.py

```diff
@@ -0,0 +1,160 @@
+
+    # length of the redeem script
+    pubkeys = multisig_get_pubkeys(multisig)
+    redeem_script_length = scripts.output_script_multisig_length(pubkeys, multisig.m)
+
+    # length of the result
+    total_length = 0
+    for s in signatures:
+        total_length += 1 + len(s) + 1  # length, signature, hash_type
+    total_length += 1 + redeem_script_length  # length, script
+
+    w = empty_bytearray(total_length)
+
+    for s in signatures:
+        if len(s):
+            scripts.append_signature(w, s, hash_type)
+
+    # redeem script
+    write_op_push(w, redeem_script_length)
+    scripts.write_output_script_multisig(w, pubkeys, multisig.m)
+
+    return w
+
+
+# A submission for an address hash.
+def output_script_sstxsubmissionpkh(addr: str) -> bytearray:
+    try:
+        raw_address = base58.decode_check(addr, blake256d_32)
+    except ValueError:
+        raise wire.DataError("Invalid address")
+    w = empty_bytearray(26)
+    w.append(0xBA)  # OP_SSTX
+    w.append(0x76)  # OP_DUP
+    w.append(0xA9)  # OP_HASH160
+    w.append(0x14)  # OP_DATA_20
+    w.extend(raw_address[2:])
+    w.append(0x88)  # OP_EQUALVERIFY
+    w.append(0xAC)  # OP_CHECKSIG
+    return w
+
+
+# A currently unused stake change script. Output amount has been checked to be
+# zero. The addr is also checked as to whether it pays to a zeroed hash.
+def output_script_sstxchange(addr: str) -> bytearray:
+    try:
+        raw_address = base58.decode_check(addr, blake256d_32)
+    except ValueError:
+        raise wire.DataError("Invalid address")
+    # Using change addresses is no longer common practice. Inputs are split
+    # beforehand.
+    for b in raw_address[2:]:
+        if b != 0:
+            raise wire.DataError("Only zeroed addresses accepted for sstx change")
+    w = empty_bytearray(26)
+    w.append(0xBD)  # OP_SSTXCHANGE
+    w.append(0x76)  # OP_DUP
+    w.append(0xA9)  # OP_HASH160
+    w.append(0x14)  # OP_DATA_20
+    w.extend(raw_address[2:])
+    w.append(0x88)  # OP_EQUALVERIFY
+    w.append(0xAC)  # OP_CHECKSIG
+    return w
+
+
+# Spend from a stake revocation.
+def input_script_ssrtx(pkh: bytes) -> bytearray:
+    utils.ensure(len(pkh) == 20)
+    s = bytearray(26)
+    s[0] = 0xBC  # OP_SSRTX
+    s[1] = 0x76  # OP_DUP
+    s[2] = 0xA9  # OP_HASH160
+    s[3] = 0x14  # OP_DATA_20
+    s[4:24] = pkh
+    s[24] = 0x88  # OP_EQUALVERIFY
+    s[25] = 0xAC  # OP_CHECKSIG
+    return s
+
+
+# Spend from a stake generation.
+def input_script_ssgen(pkh: bytes) -> bytearray:
+    utils.ensure(len(pkh) == 20)
+    s = bytearray(26)
+    s[0] = 0xBB  # OP_SSGEN
+    s[1] = 0x76  # OP_DUP
+    s[2] = 0xA9  # OP_HASH160
+    s[3] = 0x14  # OP_DATA_20
+    s[4:24] = pkh
+    s[24] = 0x88  # OP_EQUALVERIFY
+    s[25] = 0xAC  # OP_CHECKSIG
+    return s
+
+
+# Retrieve pkh bytes from a stake commitment OPRETURN.
+def pkh_from_sstxcommitment(s: bytes) -> bytes:
+    utils.ensure(len(s) == 30)
+    pkh = s[:20]
+    return pkh
```

Comment:
>     * What are the remaining 10 bytes? If there is some documentation of this, then please post a link.

Amount and fee limit flags in little endian. 

description: https://github.com/decred/dcrd/blob/b33b46cf1a24e8555e64167743c736beb31720c5/blockchain/validate.go#L70-L83

creating: https://github.com/decred/dcrwallet/blob/dfc4c75013349847235903c34231507820c8bef0/wallet/addresses.go#L347-L355

the last "58": https://github.com/decred/dcrwallet/blob/dfc4c75013349847235903c34231507820c8bef0/wallet/createtx.go#L39-L41

consensus value check: https://github.com/decred/dcrd/blob/b33b46cf1a24e8555e64167743c736beb31720c5/blockchain/validate.go#L1953-L2017

>     * Shouldn't these 10 bytes be checked too?

Ok. I will add this check.

>     * From an example transaction I can see that the first 8 of the 10 bytes encode the total input amount to the transaction. Is that always the case? If so, then why is this data duplicated?

I'm sorry, I don't quite follow. If you mean why the amount in the commitment script, it is because they were meant to be broken up at first, but we don't mess with that in practice any longer.

> What if there is a second commitment, do they have to add up to the input amount?

Yes. But for the purposes of this pr, we will only allow one.

> 
>     * What about the last 2 bytes? From the example transaction I can see they are "\x00\x58". What do they signify?

"Fee limits." The first byte is for votes, the second for revocations. Relevant code here: https://github.com/decred/dcrd/blob/master/blockchain/validate.go#L2229-L2296

It's probably fine for the foreseeable to ensure it is the default "0x0058"

> 
>     * I understand that the network checks the correctness of the amount in the OP_RETURN data, but then why does this information need to be signed?


Sorry, I don't quite follow again. I believe `SIGHASH_ALL` requires all output scripts when signing.